lib/transport/tls-session.c

```diff
@@ -195,11 +198,30 @@ tls_session_verify(TLSSession *self, int ok, X509_STORE_CTX *ctx)
 
   int ctx_error_depth = X509_STORE_CTX_get_error_depth(ctx);
   /* accept certificate if its fingerprint matches, again regardless whether x509 certificate validation was successful */
-  if (ok && ctx_error_depth == 0 && !tls_session_verify_fingerprint(ctx))
+
+  if (ctx_error_depth == 0 && self->ctx->trusted_fingerprint_list)
     {
-      msg_notice("Certificate valid, but fingerprint constraints were not met, rejecting",
-                 tls_context_format_location_tag(self->ctx));
-      return 0;
+      /* trusted-keys() is present */
+      if (ok)
+        {
+          /* this is an extra constraint */
+          if (!tls_session_verify_fingerprint(ctx))
+            {
+              msg_notice("Certificate valid, but fingerprint constraints were not met, rejecting",
+                         tls_context_format_location_tag(self->ctx));
+              return 0;
+            }
+        }
+      else
+        {
+          /* this is a trust anchor that forces the key to be valid */
+          if (tls_session_verify_fingerprint(ctx))
+            {
+              msg_notice("Certificate accepted due to being present on trusted-keys()",
```

Comment:
I think we should not change the behavior of ` trusted-keys()` because, from the perspective of a user who completely understands the old behavior and just upgrades to AxoSyslog, it would be considered a security issue.

I can see that forcibly accepting an otherwise "invalid" key based on its fingerprint can be a real use case (I have some doubts about the validity of such a need), but in that case, I think it should be introduced as a new option.
I understand that the name `trusted-keys()` would be perfect for such a new option, but I think we missed this opportunity years ago.

Changing the behavior based on config-version is another option, but I wouldn't do that either because of the security nature of this matter.

Reply:
I was not completely right here: if the fingerprint alg is chosen well, it is suitable to replace the CA verification, and that can be a perfectly valid use case.

With this new implementation, we should not default to SHA1, and we should consider not letting users set a weak hash function.
To avoid any confusion, I'd remove the old option and introduce a new one, for example:

```
trusted-fingerprints(sha1, "aa:bb:cc", ...)
```

Reply:
I've dropped this patch from the series and will fold it into #137